### PR TITLE
merkleeyes to iavl

### PR DESCRIPTION
**Pull request by ethanfrey**
Rip out all traces of merkleeyes.
Use iavl.VersionedTree with optional persistence for all operations.

This lets us use the dummy app with new code for test suites, among other things.

### CODE REVIEW

File: tests/test_cli/test.sh [hunk 1]
```diff
@@ -15,6 +15,7 @@ function testExample() {
 
 	echo "Example $N"
 	$APP &> /dev/null &
+    # $APP &> ./app.out &
```
Comment (melekes) [inline tests/test_cli/test.sh]: m?

File: example/dummy/dummy.go [hunk 2]
```diff
@@ -32,6 +34,7 @@ func (app *DummyApplication) DeliverTx(tx []byte) types.Result {
 	} else {
 		app.state.Set(tx, tx)
 	}
+	fmt.Println("set data")
```
Comment (melekes) [inline example/dummy/dummy.go]: did you use it for debug or we need this?

File: example/dummy/dummy.go [hunk 3]
```diff
@@ -40,28 +43,45 @@ func (app *DummyApplication) CheckTx(tx []byte) types.Result {
 }
 
 func (app *DummyApplication) Commit() types.Result {
-	hash := app.state.Hash()
+	// Save a new version
+	var hash []byte
+	var err error
+
+	if app.state.Size() > 0 {
```
Comment (melekes) [inline example/dummy/dummy.go]: do we need this condition? I mean, by the time we commit the size should always be > 0, no?

File: example/dummy/dummy.go [hunk 4]
```diff
@@ -40,28 +43,45 @@ func (app *DummyApplication) CheckTx(tx []byte) types.Result {
...
+	}
+
 	return types.NewResultOK(hash, "")
 }
 
 func (app *DummyApplication) Query(reqQuery types.RequestQuery) (resQuery types.ResponseQuery) {
 	if reqQuery.Prove {
-		value, proof, exists := app.state.Proof(reqQuery.Data)
+		value, proof, err := app.state.GetWithProof(reqQuery.Data)
+		// be stupid here
```
Comment (melekes) [inline example/dummy/dummy.go]: I'd prefer `if this wasn't a dummy app, we'd do something smarter` here too

File: example/dummy/dummy.go [hunk 5]
```diff
@@ -40,28 +43,45 @@ func (app *DummyApplication) CheckTx(tx []byte) types.Result {
...
+		if err != nil {
+			panic(err)
+		}
 		resQuery.Index = -1 // TODO make Proof return index
 		resQuery.Key = reqQuery.Data
 		resQuery.Value = value
-		resQuery.Proof = proof
-		if exists {
+		resQuery.Proof = wire.BinaryBytes(proof)
+		if value == nil {
```
Comment (melekes) [inline example/dummy/dummy.go]: shouldn't this be `if value != nil { "exists"`

File: example/dummy/persistent_dummy.go [hunk 6]
```diff
@@ -86,18 +89,22 @@ func (app *PersistentDummyApplication) CheckTx(tx []byte) types.Result {
...
-		Height:  app.blockHeader.Height,
-		AppHash: appHash, // this hash will be in the next block header
-	}
+	// Save a new version
+	var appHash []byte
+	var err error
 
-	app.logger.Info("Saving block", "height", lastBlock.Height, "root", lastBlock.AppHash)
-	SaveLastBlock(app.db, lastBlock)
+	if app.app.state.Size() > 0 {
```
Comment (melekes) [inline example/dummy/persistent_dummy.go]: same comment as above

File: example/dummy/persistent_dummy.go [hunk 7]
```diff
@@ -86,18 +89,22 @@ func (app *PersistentDummyApplication) CheckTx(tx []byte) types.Result {
 }
 
 func (app *PersistentDummyApplication) Commit() types.Result {
-	// Save
-	appHash := app.app.state.Save()
-	app.logger.Info("Saved state", "root", appHash)
+	app.height = app.blockHeader.Height
```
Comment (melekes) [inline example/dummy/persistent_dummy.go]: shouldn't we do this after we successfully saved a new version?

File: example/dummy/dummy.go (hunk 2 above)
Comment (ethanfrey) [reply example/dummy/dummy.go]: Debug. Oops

File: example/dummy/dummy.go (hunk 3 above)
Comment (ethanfrey) [reply example/dummy/dummy.go]: Well. We have tests that commit a few empty blocks before adding a TX. They were returning errors here, which broke things. So I added tests. Thus the slack ping to alexis

File: example/dummy/dummy.go (hunk 5 above)
Comment (ethanfrey) [reply example/dummy/dummy.go]: Yes. Good eye

File: example/dummy/persistent_dummy.go (hunk 7 above)
Comment (ethanfrey) [reply example/dummy/persistent_dummy.go]: I guess so, but if save fails, we panic, and this is only stored locally, so it doesn't really matter.
But sure, I'll clean it up.

File: example/dummy/dummy.go [hunk 8]
```diff
@@ -40,28 +41,45 @@ func (app *DummyApplication) CheckTx(tx []byte) types.Result {
...
 	return types.NewResultOK(hash, "")
 }
 
 func (app *DummyApplication) Query(reqQuery types.RequestQuery) (resQuery types.ResponseQuery) {
 	if reqQuery.Prove {
-		value, proof, exists := app.state.Proof(reqQuery.Data)
+		value, proof, err := app.state.GetWithProof(reqQuery.Data)
+		// if this wasn't a dummy app, we'd do something smarter
+		if err != nil {
+			panic(err)
```
Comment (ebuchman) [inline example/dummy/dummy.go]: does this panic if not found

File: example/dummy/dummy.go (hunk 3 above)
Comment (ebuchman) [reply example/dummy/dummy.go]: Hm. For the persistent dummy I ended up adding the constraint that InitChain must be called, which will make the state non-zero. In general would be nice if we could still get versions of empty trees so we dont have to worry about this edge condition

Comment (cloudhead) [reply example/dummy/dummy.go]: We can use `!app.state.IsEmpty()` now which is a bit cleaner, but if a tree has no root, what are we saving exactly? We could potentially have it just save a version number but hmmm..

File: example/dummy/persistent_dummy.go [hunk 9]
```diff
@@ -36,17 +35,22 @@ type PersistentDummyApplication struct {
...
-	stateTree := iavl.NewIAVLTree(0, db)
-	stateTree.Load(lastBlock.AppHash)
+	db, err := dbm.NewGoLevelDB(name, dbDir)
+	if err != nil {
+		panic(err)
+	}
 
-	// log.Notice("Loaded state", "block", lastBlock.Height, "root", stateTree.Hash())
+	stateTree := iavl.NewVersionedTree(500, db)
+	if !empty {
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply example/dummy/persistent_dummy.go]: yes, this is possible now, with fixes by alexis. I will update to the newest iavl develop and then remove this check.

File: example/dummy/persistent_dummy.go [hunk 10]
```diff
@@ -57,9 +61,8 @@ func (app *PersistentDummyApplication) SetLogger(l log.Logger) {
 
 func (app *PersistentDummyApplication) Info(req types.RequestInfo) (resInfo types.ResponseInfo) {
 	resInfo = app.app.Info(req)
-	lastBlock := LoadLastBlock(app.db)
-	resInfo.LastBlockHeight = lastBlock.Height
-	resInfo.LastBlockAppHash = lastBlock.AppHash
+	resInfo.LastBlockHeight = app.height
+	resInfo.LastBlockAppHash = app.app.state.Hash()
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply example/dummy/persistent_dummy.go]: app.state.Hash() is the root hash of the last committed tree, given the new versionedtree functionality. which was the same as lastBlock, no?

File: example/dummy/dummy.go (hunk 3 above)
Comment (ebuchman) [reply example/dummy/dummy.go]: It's just for useability. The tree is still something the app thinks about and has a root (even if empty). And for recover logic, we like the tree version to equal the height, even if the tree hasn't changed.

**Pull request merged**